CIF: --separate-chains --separate-models implementations

danny305 opened the pull request:
I got both `--separate-chains` and  `--separate-models` working. I have tested it with a handful of PDBs so far. I essentially used your implementation for the pdb format (from structure.c) as my outline. 

What has not yet been implemented yet is the n_chain_groups condition (lines 306 - 325). I refactored this code block a bit but I do not have a cpp/cif version for `freesias_structure_get_chains`. 

I am unable to get the linker to find `int freesasa_warn(const char *format, ...)` from the util.c file. I am linking with libfreesasa.a so I do not know what is going on here. I even put the declaration in cif.hh. (I try using it in cif.cc line: 253)

My implementations clearly do not test all the possible edge cases and misallocation of memory  like your original source code does. Mainly because I have never really thought about that stuff so I would like for you to take lead on these details.  I want to get this from dev to production ready and learn from you as we do it.

mittinatten commented:
Regarding linking `freesasa_warn` it should be solved if you add 
```c
#ifdef __cplusplus
extern "C" {
#endif
```
near the top of `freesasa_internal.h` and
```c
#ifdef __cplusplus
}
#endif
```
near the end. (You can compare with the `freesasa.h` to see where it should be placed.) 

This tells the C++ compiler that the functions listed are from a C library and should be linked differently (don't ask me about the details 😃 ).

mittinatten commented:
Very nice that you got this far already! I'll do a thorough review soon. At a first glance there are two things though I wanted to mention. 
1. From the Travis logs (you can see them if you click the red cross next to the last commit) I see there is a seg-fault for `--chain-groups`. 
2. We should coordinate our indentation and brace styles, there is a `.clang-format` in the repo that the rest of the code follows. Can you apply that to your code too? (I have vscode setup to do it automatically on save)

danny305 commented:
Yeah the `--chain-groups` is what I am referring to in my second point ^. (I do not have a cpp/cif version for `freesias_structure_get_chains`).

Its on the to-do list. I ill update the target branch probably today/tomorrow. 

Ill give your extern C thing a go. Are you sure its the freesias_internal.h? the definition is in util.c. Can you elaborate?

mittinatten commented:
Right, so when you compile the C++ code you just include the C header, the `extern "C"` declarations tells the compiler that these functions should be flagged for different linkage later when the library is linked in. The C++ compiler never touches util.c it just uses the declarations given in the header. (At least that's my understanding of it.)

I got the chain-groups to work with cif before, by just mapping to freesasa_structure and then reusing the old code, but leveraging the gemmi tools directly is probably a better solution in the long run. Especially if we anticipate a migration to gemmi also for PDB files eventually.

danny305 commented:
Simon. I did as you suggested to the `freesias_internal.h` file and it was able to link the function `freesasa_warn()`.

I also fixed the memory error in --chain-groups. It was a typo on my end.

Next I will look into syncing up our indentation and brace style. 

What else do we need to do for this PR? to merge the feature/branch to master?

mittinatten commented:
Great! I will try to find time for a code review today, and then we can probably merge this PR.

There are at least two more things we need to do. 
1. Run the larger test set. I am getting close to having a runner ready, can probably wrap that up this week.
2. Documentation. We should add the new feature to the man pages, the help message in the CLI and the online docs. Shouldn't be necessary to have more than one or two sentences either place.

danny305 commented:
Okay I'll get that done tomorrow. 

I can add a txt file to the repo tomorrow with several thousand protein PDB IDs and a python/shell script to pull down both the CIF and pdb files for a given protein. Do you have a preference if I use bash vs Python? Where would you like me to put this? I was thinking rootDir/dataset. Let me know if you prefer it else where. 

What else can I do?

mittinatten commented:
I have a script almost ready that does the downloading and test running, I just need to polish the error reporting a bit, so all I need is the list of PDB IDs

mittinatten commented:
I learning F# at work right now, and running a sequence of commands and summing up the errors overlaps with a lot of the concepts we're discussing at the moment. So I decided to write the test runner in F# as an exercise, which means it takes a bit longer to write, but I hope it will result in pretty neat code.

danny305 commented:
I'll get this to you in the morning. I'll just put the txt file in the root directory. 

I'm going to spend sometime learning getopt and start working on editing the CLI for outputting a cif. 

I'll also try to get cmake to build config.h based on the local system of the user.

I'll slowly implement as much as I can for the cmake build pipeline for whenever we are ready for that feature.

danny305 commented:
Wow! I've never heard of F#. Sounds super cool. I'll get you that txt file here in the next few hours.

mittinatten commented on src/cif.cc:
```diff
@@ -60,12 +82,61 @@ structure_from_doc(const gemmi::cif::Document &doc,
-            auto model = atoi(site[11].c_str());
+            if (models.count(std::stoi(site[11])) == 0) continue;
+            
+            freesasa_cif_atom atom = {
+                .group_PDB = site[0].c_str(),
+                .auth_asym_id = site[1][0],
+                .auth_seq_id = site[2].c_str(),
+                .pdbx_PDB_ins_code = site[3].c_str(),
+                .auth_comp_id = site[4].c_str(),
+                .auth_atom_id = site[5].c_str(),
+                .label_alt_id = site[6].c_str(),
+                .type_symbol = site[7].c_str(),
+                .Cartn_x = atof(site[8].c_str()),
+                .Cartn_y = atof(site[9].c_str()),
+                .Cartn_z = atof(site[10].c_str())};
```
This could perhaps be a separate function, e.g. `atom_from_site()`?

danny305 replied:
Yeah I agree. I can do it.

mittinatten commented on src/cif.cc:
```diff
@@ -60,12 +82,61 @@ structure_from_doc(const gemmi::cif::Document &doc,
+                .Cartn_y = atof(site[9].c_str()),
+                .Cartn_z = atof(site[10].c_str())};
+
+            auto currentAltId = site[6][0];
+
+            if (!(structure_options & FREESASA_INCLUDE_HYDROGEN) && std::string(atom.type_symbol) == "H") {
+                continue;
+            }
+
+            // Pick the first alternative conformation for an atom
+            if (currentAltId != '.' && currentAltId != 'A') {
+                continue;
+            }
+
+            freesasa_structure_add_cif_atom(structure, &atom, classifier, structure_options);
```
There's some repetition here, too, but not sure it's worth it to extract it to a function.

mittinatten commented on src/cif.cc:
```diff
@@ -60,12 +82,61 @@ structure_from_doc(const gemmi::cif::Document &doc,
+static freesasa_structure*
+structure_from_doc(const gemmi::cif::Document &doc,
+                         const std::string &filter,
+                         const int &col, 
```
There's no benefit in this being a reference, right?

danny305 replied:
Are you talking about the col or the filter? The filter is a string and I didn't want to copy it. There is no benefit to referencing the int.

mittinatten commented on src/cif.cc:
```diff
@@ -60,12 +82,61 @@ structure_from_doc(const gemmi::cif::Document &doc,
+                         const freesasa_classifier *classifier,
+                         int structure_options)
+{
+    std::cout << "Creating a filtered freesasa structure !" << std::endl;
```
Only strictly necessary things should be in the output I think. So I think the `cout`s should be deleted or inside `#ifdef DEBUG`

danny305 replied:
Sorry I meant to delete all these! I missed one. good catch. 
Ill start using `#DEFINE DEBUG`.

mittinatten commented on src/cif.cc:
```diff
@@ -60,12 +82,61 @@ structure_from_doc(const gemmi::cif::Document &doc,
                 continue;
             }
 
+            if (filter != site[col]) continue;
```
This was an efficient solution! (But perhaps the variable should have a slightly more descriptive name, maybe `filter_col` or something like that?)

danny305 replied:
Yeah that's a good point.

mittinatten commented on src/cif.cc:
```diff
@@ -96,24 +167,117 @@ structure_from_doc(const gemmi::cif::Document &doc,
+            int n_new_chains  = chain_names->size();
+            n_chains         +=  n_new_chains;
+
+            if (n_new_chains == FREESASA_FAIL) gemmi::fail("No chains in protein");
```
Should probably use `freesasa_fail()` here, so the error message looks the same as the other errors.

mittinatten replied:
Also I think this can never happen, right? `FREESASA_FAIL == -1`

danny305 replied:
Yeah this was before I got `freesasa_warn()` working. Ill refactor.

mittinatten commented on src/cif.cc:
```diff
@@ -96,24 +167,117 @@ structure_from_doc(const gemmi::cif::Document &doc,
+    /* for each model read chains if requested */
+    if (options & FREESASA_SEPARATE_CHAINS) 
+    {
+        for (auto& model : models) 
```
Am I missing something, or will this give the wrong result if we have `FREESASA_SEPARATE_MODELS`?

mittinatten reviewed (changes requested):
This is a good start, and looks like most things are in place. I added a few questions/suggestions in the code, that might not require any changes. Before we merge these are the things that are missing as far as I can tell:

- [ ] Check if separate-chains and separate-models work in combination (from the look of the code, they might not)
- [ ] Remove log messages
- [ ] Remove commented out code
- [ ] Apply correct styling, as mentioned
- [ ] We should have a few CLI tests that check that we get the same results for these options as for PDB files. (can be added here: https://github.com/mittinatten/freesasa/blob/feature/cif/tests/test-cli.in#L222) Let me know if I can help with that.

I really appreciate the work your putting in to this!

mittinatten commented:
By the way, I'm not sure about cmake, the autotools setup works, and has the advantage that everything needed will be installed on most systems. It is a bit of a pain to maintain, but rarely needs to be changed. So don't go too far down that road before we've discussed it further. We can make it a separate issue anyway.

danny305 commented:
On the CMake point, I am just in the process of learning cmake. So I have been using this as an opportunity to learn. There is no need to add it if its not needed. I just thought I would keep you in the loop on that front.

danny305 commented:
So I addressed all the issues besides the last one. I added a few pdb/cif pairs to the tests/data directory. If you could help turn these into tests that would be great. I have manually compared to the cif/pdb pairs and they pass the eye test. 

I don't know why the travis stuff is failing. The errors do not make sense to me. Maybe it something on your end? 

Im going to make 1 more commit tonight where I add a txt file of diverse pdb codes.

mittinatten commented on src/cif.cc:
```diff
@@ -60,24 +91,47 @@ structure_from_doc(const gemmi::cif::Document &doc,
+    freesasa_structure *structure = freesasa_structure_new();
+
+    for (auto block : doc.blocks) {
+        auto prevAltId = '?';
```
this isn't needed anymore right? Should have mentioned in the last review

mittinatten replied:
(same in the other variant)

danny305 commented:
I added the tests but for some reason travis is unable to compile the code. Could you have a look?

mittinatten commented:
Great! Just let me know when you are ready for a final review. My testrunner works now - I want to experiment a bit more with refactoring it before merging the code, for the sake of learning - but I'll run the tests before that, so we now where we stand with regards to finishing this feature.

danny305 commented:
Its ready for your approval boss.

mittinatten commented on src/cif.cc:
```diff
@@ -44,48 +110,48 @@ static const auto atom_site_columns = std::vector<std::string>({
             }
 
-            auto model = atoi(site[11].c_str());
+            if (discriminator(site)) continue;
```
nice!

mittinatten reviewed (approved):
I think the last refactoring was an elegant solution! Very nice. I really appreciate the effort you put in!
